opm/utility/ECLFluxCalc.cpp

```diff
@@ -71,16 +229,282 @@ namespace Opm
+    {
+        auto dyn_data = DynamicData{};
+
+        // Step 1 of Phase Pressure Calculation.
+        // Retrieve oil pressure directly from result set.
+        dyn_data.pressure = this->graph_
+            .linearisedCellData(rstrt, "PRESSURE",
+                                &ECLUnits::UnitSystem::pressure);
+
+        // Step 1 of Mobility Calculation.
+        // Store phase's relative permeability values.
+        dyn_data.mobility =
+            this->satfunc_.relperm(this->graph_, rstrt, phase);
+
+        // Step 1 of Mass Density (Reservoir Conditions) Calculation.
+        // Allocate space for storing the cell values.
+        dyn_data.density.assign(this->graph_.numCells(), 0.0);
+
+        switch (phase) {
+        case ECLPhaseIndex::Aqua:
+            return this->watPVT(std::move(dyn_data));
+
+        case ECLPhaseIndex::Liquid:
+            return this->oilPVT(rstrt, std::move(dyn_data));
+
+        case ECLPhaseIndex::Vapour:
+            return this->gasPVT(rstrt, std::move(dyn_data));
+        }
+
+        throw std::invalid_argument {
+            "phaseProperties(): Invalid Phase Identifier"
+        };
+    }
+
+
+
+
+
+    ECLFluxCalc::DynamicData
+    ECLFluxCalc::gasPVT(const ECLRestartData& rstrt,
+                        DynamicData&&         dyn_data) const
+    {
+        verify_active_phase(this->pvtGas_, "Gas");
+
+        const auto rv = vapoilVector(this->graph_, this->vapoil_, rstrt);
+
+        this->regionLoop([this, &rv, &dyn_data]
+            (const int regID)
+        {
+            // Note: This function assumes that 'regID' is a traditional
+            // ECL-style one-based region ID such as PVTNUM.  Subtract one,
+            // where approriate, to generate zero-based region indices.
+
+            const auto Rv = ECLPVT::Gas::VaporizedOil {
+                this->gatherRegionSubset(regID, rv)
+            };
+
+            const auto Pg = ECLPVT::Gas::GasPressure {
+                // Cheating.  This is Po.
+                this->gatherRegionSubset(regID, dyn_data.pressure)
+            };
+
+            // Mass Density at Reservoir Conditions.  Relies on setup code
+            // having allocated sufficient space.
+            {
+                const auto rhoOS = this->vapoil_
+                    ? this->pvtOil_->surfaceMassDensity(regID - 1)
+                    : 0.0;
+
+                const auto rhoGS =
+                    this->pvtGas_->surfaceMassDensity(regID - 1);
+
+                const auto Bg = this->pvtGas_
+                    ->formationVolumeFactor(regID - 1, Rv, Pg);
+
+                auto rhoGr = std::vector<double>{};
+                rhoGr.reserve(Bg.size());
+
+                std::transform(std::begin(Bg),
+                               std::end  (Bg),
+                               std::begin(Rv.data),
+                               std::back_inserter(rhoGr),
+                    [rhoOS, rhoGS]
+                    (const double Bg, const double Rv)
```

Comment:
Again, hard to untangle and shadowing variable.

Reply:
> shadowing variable.

Not really.  We're not capturing `Bg` or `Rv` from the environment.

Reply:
No, but there is some potential for confusion, naming the vector and the element the same.